### PR TITLE
[mypyc] Add support for building mypyc code on WASM

### PR DESCRIPTION
**Pull request by ethanhs**
### Description

This PR tweaks two things about how mypyc generates and builds C code to better support WebAssembly. First, we search `sysconfig` for the size of `size_t`, which works much better for cross-compiling. Second, newer versions of clang have `-Wno-unused-but-set-variable` and so it is added to the default list of arguments (this should probably land regardless the decision on merging this PR).

## Test Plan

This PR depends on https://github.com/python/mypy/pull/13445. To test this PR, you can do the following:

*assuming mypy checkout with both PRs applied, must be on Python 3.10(!)*
```
$ pip install pyodide-build
$ pyodide build --exports pyinit backend-args --global-option=--use-mypyc
```
Note: you will get a warning about using `--global-option`, you can ignore it for now. I'm trying to find out why `--build-option` isn't working...

### CODE REVIEW

File: mypyc/build.py [hunk 1]
```diff
@@ -546,10 +546,8 @@ def mypycify(
             "-Wno-unused-variable",
             "-Wno-unused-command-line-argument",
             "-Wno-unknown-warning-option",
+            "-Wno-unused-but-set-variable",
```
Comment (JelleZijlstra) [inline mypyc/build.py:549]: Just to confirm, do compilers generally ignore unused -W flags?

Comment (ethanhs) [reply mypyc/build.py:549]: The config `-Wno-unknown-warning-option` on the line before should silence any errors about this I believe.

**Review by JelleZijlstra — APPROVED** [review]

File: mypyc/build.py (hunk 1 above)
Comment (ethanhs) [reply mypyc/build.py:536]: Apparently I spoke too soon :/

```
cc1: error: unrecognized command line option ‘-Wno-unknown-warning-option’
cc1: error: unrecognized command line option ‘-Wno-unused-command-line-argument’
```

**Comment by ethanhs** [comment]
Blocked on https://github.com/python/typeshed/pull/8572

**Comment by ethanhs** [comment]
Hmm I'm unsure why things disagree. It seems like `sys.maxsize != sysconfig.get_config_var("SIZEOF_SIZE_T")` on 32 bit Python on 64bit windows??

**Comment by ethanhs** [comment]
Ah yeah I messed up my math so it was setting `SIZEOF_SIZE_T` to 3 not 4 😅 

This should be ready review now!

**Comment by ethanhs** [comment]
The latest commit enables outputting VT100 color codes for use in the web via things like https://www.npmjs.com/package/ansi-to-html. I guess I could directly output HTML but I figure that this is a bit more portable in case someone wants to run mypy under node or do something else. This is only enabled if you pass `MYPY_FORCE_COLOR` anyway.

**Review by hauntsaninja — APPROVED** [review]
Looks good, with the big caveat that I'm not sure I understand the exact circumstances when sysconfig's sizeof_size_t differs from sys.maxsize

**Comment by ethanhs** [comment]
@hauntsaninja 
> I'm not sure I understand the exact circumstances when sysconfig's sizeof_size_t differs from sys.maxsize

`SIZE_OF_SIZE_T` is only defined in UNIX environments, so we still rely on `sys.maxsize` on Windows. They only differ when you are cross compiling (like we would be for WASM)

**Comment by ethanhs** [comment]
Oh this also needs https://github.com/python/mypy/pull/13445 to actually work (pyodide build uses the PEP 517 interface)

**Pull request merged**